crates/matrix-sdk-crypto/src/session_manager/group_sessions/share_strategy.rs

```diff
@@ -127,73 +152,91 @@ pub(crate) async fn collect_session_recipients(
-                // We only need the user identity if only_allow_trusted_devices is set.
-                let device_owner_identity = if only_allow_trusted_devices {
-                    store.get_user_identity(user_id).await?
-                } else {
-                    None
-                };
-                split_recipients_withhelds_for_user(
+    match settings.sharing_strategy {
+        CollectStrategy::DeviceBasedStrategy {
+            only_allow_trusted_devices,
+            error_on_unsigned_device_of_verified_users,
+        } => {
+            let mut unsigned_devices_of_verified_users: BTreeMap<OwnedUserId, Vec<OwnedDeviceId>> =
+                Default::default();
+            for user_id in users {
+                let user_devices = store.get_device_data_for_user_filtered(user_id).await?;
+
+                // We only need the user identity if only_allow_trusted_devices or
+                // error_on_unsigned_device_of_verified_users is set.
+                let device_owner_identity =
+                    if only_allow_trusted_devices || error_on_unsigned_device_of_verified_users {
+                        store.get_user_identity(user_id).await?
+                    } else {
+                        None
+                    };
+                let recipient_devices = split_recipients_withhelds_for_user(
                     user_devices,
                     &own_identity,
                     &device_owner_identity,
                     only_allow_trusted_devices,
-                )
+                    error_on_unsigned_device_of_verified_users,
+                );
+
+                let recipients = recipient_devices.allowed_devices;
+                let withheld_recipients = recipient_devices.denied_devices_with_code;
+
+                // If we haven't already concluded that the session should be
+                // rotated for other reasons, we also need to check whether any
+                // of the devices in the session got deleted or blacklisted in the
+                // meantime. If so, we should also rotate the session.
+                if !should_rotate {
+                    should_rotate = is_session_overshared_for_user(outbound, user_id, &recipients)
+                }
+
+                devices.entry(user_id.to_owned()).or_default().extend(recipients);
+                withheld_devices.extend(withheld_recipients);
+                if let Some(blockers) = recipient_devices.unsigned_of_verified_user {
+                    if !blockers.is_empty() {
+                        unsigned_devices_of_verified_users
+                            .entry(user_id.to_owned())
+                            .or_default()
+                            .extend(blockers)
+                    }
+                }
             }
-            CollectStrategy::IdentityBasedStrategy => {
+
+            if error_on_unsigned_device_of_verified_users
+                && !unsigned_devices_of_verified_users.is_empty()
+            {
```

Comment:
Yes it is. I am hesitant on it. 
By putting it, it makes it very clear for the reader that we only throw if this settings is on.
The fact that in `split_recipients_withhelds_for_user` we only put the device in the `unsigned_devices_of_verified_users`  is more an optimization to avoid checking signatures when it's not needed.

There is another similar optimization thing in this code, like for the `device_owner_identity` we only access the store if `only_allow_trusted_devices || error_on_unsigned_device_of_verified_users` so depending on the setting the identity could be some or None.

So yeah not sure what is best :/

Reply:
>  By putting it, it makes it very clear for the reader that we only throw if this settings is on.

I think we can convey this in a comment.

```suggestion
            // `unsigned_devices_of_verified_users` will only be populated if `error_on_unsigned_device_of_verified_users` was set.
            // We should return an error if there are any entries.
            if !unsigned_devices_of_verified_users.is_empty() {
```

For me, it's actually less about optimisation (the compiler can do a better job of optimizing this than I can anyway), and more about emphasising the fact that we don't need to worry about what it means if `unsigned_devices_of_verified_users` has entries but `error_on_unsigned_device_of_verified_users` is false.